### PR TITLE
contrib: Docker integration tests

### PR DESCRIPTION
**Pull request by dmcgowan**
Add a full suite of integration tests between Docker and the registry. This includes varies TLS and authentication methods as well as various push/pull scenarios.

This strategy is heavily adopted from the swarm project. The swarm project runs a suite of bats scripts inside of a Docker-in-Docker environment.

- [x] Docker-in-Docker test runner
- [x] Use bats
- [x] Deprecate and remove test_docker.sh
- [x] Reduce run time by using volume
- [x] Define test scenarios (please suggest additional test cases in comments)
- [x] Push up image to distribution/docker-integration
- [ ] Review test strategy with @stevvooe 
- [ ] LGTMs from @RichardScothern @dmp42

### CODE REVIEW

File: contrib/docker-integration/test_runner.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+cd "$(dirname "$(readlink -f "$BASH_SOURCE")")"
+
+# Load the helpers.
+#. helpers.bash
+
+TESTS=${@:-.}
+
+# Drivers to use for Docker engines the tests are going to create.
+STORAGE_DRIVER=${STORAGE_DRIVER:-overlay}
```
Comment (dmcgowan) [inline contrib/docker-integration/test_runner.sh]: I changed this from aufs locally, probably best to leave as aufs and configure my local environment appropriately.

Comment (dmcgowan) [inline contrib/docker-integration/test_runner.sh]: Second thought, will leave as overlay, added a section in the readme which mentions how to use aufs.

File: contrib/docker-integration/README.md [hunk 2]
```diff
@@ -26,15 +48,14 @@ between a TLS configuration with a CA and self-signed.
 
 		$  sudo chmod +x /usr/local/bin/docker-compose
```
Comment (dmp42) [inline contrib/docker-integration/README.md]: I think I would gladly get rid of stuff like that, and replace by a link to "Install Docker Compose" (reduces duplication, reduces maintenance).

Comment (dmcgowan) [inline contrib/docker-integration/README.md]: Agreed, updated

**Comment by dmcgowan** [comment]
Additional test cases I am creating, mostly around pull security
- Bad signature (manifest data changed)
- Renamed image (fails in Docker 1.7)
- Incorrect remote digest
- Incorrect digest on pull by digest
- Altered data blob (swapped out for different layer)

**Comment by dmcgowan** [comment]
This change is ready for further review. The additional tests I plan to add in further PRs and keep this one about the strategy and initial implementation.

File: contrib/docker-integration/tls.bats [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+# Login information, should match values in nginx/test.passwd
+user="testuser"
+password="passpassword"
+email="distribution@docker.com"
+
+function setup() {
+	docker pull $image
+}
+
+# has_digest enfoces the last output line is "Digest: sha256:..."
```
Comment (RichardScothern) [inline contrib/docker-integration/tls.bats]: s/enfoces/enforces

**Comment by dmcgowan** [comment]
Can we go ahead and get this merged. I have other tests I want to add but do not want to continuously add to this PR.

Relevant future tests
- Proxy cache
- MITM registry
- Fallback logic

**Comment by RichardScothern** [comment]
LGTM, I will make use of this.

**Comment by stevvooe** [comment]
LGTM

**Pull request merged**